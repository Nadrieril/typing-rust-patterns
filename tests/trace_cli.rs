use assert_cmd::Command;
use std::collections::hash_map::DefaultHasher;
use std::hash::{Hash, Hasher};

#[test]
fn trace_cli() -> anyhow::Result<()> {
    let test_cases = &[
        "quit",
        "help",
        "options",
        "set",
        "set mistyped",
        "set mistyped val",
        "set ref_binding_on_inherited true",
        "set allow_ref_pat_on_ref_mut false",
        "set downgrade_shared_inside_shared true",
        "set ergo2024\noptions",
        "&x: &mut T",
        "set allow_ref_pat_on_ref_mut false\n&x: &mut T",
        "rules",
    ];
    for &test_case in test_cases {
        let success = Command::cargo_bin("typing-rust-patterns")?
            .write_stdin(test_case)
            .assert()
            .success();
        let out = String::from_utf8(success.get_output().stdout.clone())?;

        // Identify each snapshot file by the hash of the test case.
        let req_hash = {
            let mut hasher = DefaultHasher::new();
            test_case.hash(&mut hasher);
            hasher.finish().to_string()
        };
        insta::with_settings!({
            snapshot_suffix => req_hash,
            info => &test_case,
            omit_expression => true,
            prepend_module_to_snapshot => true,
        }, {
            insta::assert_snapshot!(out);
        });
    }
    Ok(())
}
