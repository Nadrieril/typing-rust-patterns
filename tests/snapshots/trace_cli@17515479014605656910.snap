---
source: tests/trace_cli.rs
info: set
---
Error: Syntax is `set option value`.

Options are:
- match_constructor_through_ref: false | true
    whether `[p]` can match on `&[T]`; the heart of match ergonomics.
- eat_inherited_ref_alone: false | true
    whether `&p`/`&mut p` is allowed on an inherited reference if the underlying type isn't also a reference type
- inherited_ref_on_ref: EatOuter | EatInner | EatBoth
    how to handle a reference pattern on a double reference when the outer one is inherited
- fallback_to_outer: false | true
    whether to try again in `EatOuter` mode when a `EatBoth` or `EatInner` case has a mutability mismatch
- eat_mut_inside_shared: false | true
    in `EatInner` or `EatBoth`, `&mut p` can eat an inner `&mut T` from under a `&`
- allow_ref_pat_on_ref_mut: false | true
    whether to allow `&p: &mut T`
- downgrade_mut_inside_shared: false | true
    RFC3627 rule 3: downgrade `&mut` inherited references to `&` inside a shared deref
- ref_binding_on_inherited: Error | ResetBindingMode | AllocTemporary
    how to handle a `ref x` binding on an inherited reference
- mut_binding_on_inherited: Error | ResetBindingMode | Keep
    how to handle a `mut x` binding on an inherited reference
- simplify_deref_mut: false | true
    whether to simplify `*&mut expr`, which removes some borrow errors
- always_inspect_bm: false | true
    whether to always branch on the binding mode when computing rules. this is required for the `SequentBindingMode` style
- predicate_style: Expression | Sequent | BindingMode | SequentBindingMode | Stateless
    the style of the typing predicate; not all rulesets can be expressed in all styles, only `Expression` is compatible with all rulesets

There also exist some predefined option-bundles. Activate one with `set <bundle>`
- nadri: a reasonable proposal; like `stateless` but forbids `ref` bindings that create temporaries
- stateless: a proposal that tracks no hidden state; purely type-based
- stable_rust: emulates the behavior of current stable rust
- rfc3627: the accepted RFC3627 behavior
- rfc3627_2021: the accepted RFC3627 behavior under edition 2021
- rfc3627_breaking_only: the breaking changes for edition 2024 planned in RFC3627
- structural: purely structural matching, with no match ergonomics
- waffle: a proposal by @WaffleLapkin (excluding the proposed rule3 extension)
